src/translator/sentence_ranges.h

```diff
@@ -46,24 +49,14 @@ class Annotation {
 
 private:
   /// A flat storage for ByteRanges. Composed of word ByteRanges, extra
-  /// information in sentenceBeginIds_ to denote sentence boundary markers as
+  /// information in sentenceEndIds_ to denote sentence boundary markers as
   /// indices.
   std::vector<ByteRange> flatByteRanges_;
 
-  /// Stores indices where sentences begin
-  std::vector<size_t> sentenceBeginIds_;
-
-  /// Returns ByteRanges corresponding to beginning and end words of sentence
-  /// corresponding to sentenceIdx. This is useful in using the information to
-  /// construct a ByteRange of a sentence taking the begin from the first and
-  /// end from the second.
-  std::pair<ByteRange, ByteRange> sentenceTerminals(size_t sentenceIdx) const;
-
-  /// Returns indices of terminal (word) ByteRanges in sentenceIds_ of a
-  /// sentence corresponding to sentenceIdx. The distance can be used to compute
-  /// number of words in a sentence (numWords) and also to construct the
-  /// terminal ByteRanges (sentenceTerminals).
-  std::pair<size_t, size_t> sentenceTerminalIds(size_t sentenceIdx) const;
+  /// Stores indices onto flatByteRanges_ of where sentences end (not inclusive,
+  /// aligned with C++ half interval notions). There is a 0 marker to simplify
+  /// sources, indicating where the -1-th sentence ends.
+  std::vector<size_t> sentenceEndIds_;
```

Comment:
Please correct me if I am wrong anywhere below. I am taking a super simple example to understand the Annotation class.

Imagine a text containing 2 sentences: `"Ab cd. Ef gh."`
So in bytes, it is:
| Bytes      | 0   | 1   | 2   | 3   | 4   | 5   | 6   | 7   | 8   | 9   | 10  | 11  | 12  |
|------------|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|
| Characters | "A" | "b" | " " | "c" | "d" | "." | " " | "E" | "f" | " " | "g" | "h" | "." |

And **assuming** the sentencepiece tokenized the individual sentences as:
Sentence 1 ("Ab cd.") = `"Ab_",    "cd",    "."`
Sentence 2 ("Ef gh.")= `"Ef_",    "gh",    "."`

(I assumed that the white space at byte 6 doesn't belong to any sentence when it comes to tokenization by sentencepiece because sentencepiece deals with only sentences. But I may be wrong.)

So `flatByteRanges_ = [(0,3), (3,5), (5,6), (7,10), (10,12), (12,13)]`

and `sentenceEndIds_ = [0, 6, 13]` which means

Sentence 1 = `(0, 6)`
Sentence 2 = `(6, 13)`

Am I right?

(Of course, all the ranges written above as `(x, x)` means half open interval `[x, x)`)

Reply:
I'll try to help you here first conceptually, through composition.

```
typedef ByteRange Word;
typedef vector<Word> Sentence; // vector<ByteRange>, a single sentence
typedef vector<Sentence> AnnotatedText;   // vector<vector<ByteRange>, multiple sentences

AnnotatedText example; 
```

How do I get a sentence ByteRange, for the sentence at index `i`? 
* `(ByteRange){example[i].front().begin, example[i].last().end}`

Annotation is the same thing, except the `vector<vector` hammered into a flat structure to avoid multiple reallocs. This is done by having markers of indices where sentence ends in `flatByteRanges_` in addition.

> sentenceEndIds_ = [0, 6, 13]

You seem to be looking at them as end indices. No, they're indices pointing to the `flatByteRanges` structure. `sentenceEndIds_ = [2, 5]`. 

This additional indirection helps when you want to return `vector<Word>` for a sentence.

Reply:
There's an extra sentence marker to convert ifs into accessors at the moment, but the above should be simpler to understand. To the outside world, this container behaves just like `vector<vector<ByteRange>` with multiple sentences.


```
sentenceEndIds_ = [2, 5] => #sentences = [2, 5].size() = 2
```

Reply:
Thanks for explanation. This makes sense to me 👍 

> they're indices pointing to the `flatByteRanges` structure.

Got it.

> `sentenceEndIds_ = [2, 5]`.

I guess, it would be equal to `[0, 2, 5]` (because you are doing `sentenceEndIds_.push_back(0)` in the constructor)? Right? It doesn't matter from the public methods perspective but asking for the sake of correctness 👍

Reply:
As of now, yes, I think your understanding is correct. numSentences is adjusted to return size of this container  - 1 accordingly to fix the external API.